src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+-- | A general-purpose fence pointer index.
+module Database.LSMTree.Internal.IndexOrdinary
+(
+    IndexOrdinary (IndexOrdinary),
+
+    -- * Search
+    search,
+
+    -- * Deserialisation
+    fromSBS
+)
+where
+
+import Prelude hiding (length, drop, takeWhile)
+
+import Control.Monad
+       (
+           when
+       )
+import Data.Word
+       (
+           Word8,
+           Word16,
+           Word32,
+           Word64,
+           byteSwap32
+       )
+import Data.Primitive.ByteArray
+       (
+           ByteArray,
+           indexByteArray
+       )
+import Data.ByteString.Short
+       (
+           ShortByteString (ShortByteString)
+       )
+import Data.ByteString.Short qualified as ShortByteString
+       (
+           length
+       )
+import Data.Vector
+       (
+           Vector,
+           fromList,
+           length,
+           (!),
+           drop,
+           takeWhile,
+           findIndexR
+       )
+import Data.Vector.Primitive qualified as Primitive
+       (
+           Vector (Vector),
+           force,
+           null,
+           length,
+           take,
+           drop,
+           splitAt,
+       )
+import Database.LSMTree.Internal.Serialise
+       (
+           SerialisedKey (SerialisedKey')
+       )
+import Database.LSMTree.Internal.Entry
+       (
+           NumEntries (NumEntries)
+       )
+import Database.LSMTree.Internal.IndexCompact
+       (
+           PageNo (PageNo),
+           PageSpan (PageSpan)
+       )
+
+{-|
+    The version of the index and its serialisation format that is supported by
+    this module.
+-}
+supportedVersion :: Word32
+supportedVersion = 1
+
+{-|
+    A general-purpose fence pointer index.
+
+    An index is represented by a vector that maps the number of each page to the
+    key stored last in this page or, if the page is an overflow page, to the key
+    of the corresponding key–value pair. From this, the following constraints on
+    the vector follow:
+
+      * The vector must be non-empty.
+
+      * The elements of the vector must be non-decreasing.
+-}
+newtype IndexOrdinary = IndexOrdinary (Vector SerialisedKey)
+
+-- * Search
+
+{-|
+    Returns the span of pages that contain the key–value pair of a certain key
+    if there is such a pair, or an arbitrary, valid page span if there is not.
```

Comment:
Nitpicking is welcome: I like precision. 🙂

That said, I think that my specification is correct, since it only says that a page span containing the search key is returned *if there is such a pair*. It doesn’t promised that the returned page span contains the key, since it could also have been returned because of the second part of the result specification: “an arbitrary, valid page span if there is not [an entry with the search key]”.

Reply:
Okay, then I think the wording confuses me. It would probably be useful to have a unified specification of the `search` result, and I think we can find a middle ground between the description of this `search` function and the one for the compact index. In particular, I think it's clearer if we mention something along the lines of these two statements:

* `search` always returns a valid page span.
* `search` may return false positives, but not false negatives.

Reply:
An improved version of the documentation is in 7bb67533b2b4c2cf853aee387f20e930e4ef7b7b. Can you please check and tell me what you think about it?